File: php/src/Algorithms/QuickSort.php
<?php

declare(strict_types=1);

namespace DSA\Algorithms;

/**
 * Quick sort algorithm with median-of-three pivot selection.
 * Time: O(n log n) average, O(n^2) worst case
 * Space: O(log n) average for recursion stack
 * Stable: No
 */
class QuickSort
{
    /**
     * Sort an array in-place using quick sort.
     *
     * @param array &$arr Array to sort (modified in-place)
     * @param callable|null $comparator Custom comparison function
     */
    public static function sort(array &$arr, ?callable $comparator = null): void
    {
        $comparator = $comparator ?? fn($a, $b) => $a <=> $b;
        self::quickSort($arr, 0, count($arr) - 1, $comparator);
    }

    /**
     * Return a sorted copy of the array using quick sort.
     *
     * @param array $arr Array to sort
     * @param callable|null $comparator Custom comparison function
     * @return array Sorted array
     */
    public static function sorted(array $arr, ?callable $comparator = null): array
    {
        self::sort($arr, $comparator);
        return $arr;
    }

    private static function quickSort(array &$arr, int $low, int $high, callable $comparator): void
    {
        if ($low < $high) {
            $pivotIndex = self::partition($arr, $low, $high, $comparator);
            self::quickSort($arr, $low, $pivotIndex - 1, $comparator);
            self::quickSort($arr, $pivotIndex + 1, $high, $comparator);
        }
    }

    private static function partition(array &$arr, int $low, int $high, callable $comparator): int
    {
        // Median-of-three pivot selection
        $mid = $low + intdiv($high - $low, 2);

        if ($comparator($arr[$mid], $arr[$low]) < 0) {
            self::swap($arr, $low, $mid);
        }
        if ($comparator($arr[$high], $arr[$low]) < 0) {
            self::swap($arr, $low, $high);
        }
        if ($comparator($arr[$mid], $arr[$high]) < 0) {
            self::swap($arr, $mid, $high);
        }

        $pivot = $arr[$high];
        $i = $low - 1;

        for ($j = $low; $j < $high; $j++) {
            if ($comparator($arr[$j], $pivot) <= 0) {
                $i++;
                self::swap($arr, $i, $j);
            }
        }

        self::swap($arr, $i + 1, $high);
        return $i + 1;
    }

    private static function swap(array &$arr, int $i, int $j): void
    {
        [$arr[$i], $arr[$j]] = [$arr[$j], $arr[$i]];
    }
}

File: ruby/lib/algorithms/quick_sort.rb
# frozen_string_literal: true

module DSA
  module Algorithms
    # Quick sort algorithm with median-of-three pivot selection.
    # Time: O(n log n) average, O(n²) worst case
    # Space: O(log n) average for recursion stack
    # Stable: No
    module QuickSort
      class << self
        # Sort an array in-place.
        def sort!(arr, &comparator)
          comparator ||= ->(a, b) { a <=> b }
          quick_sort(arr, 0, arr.size - 1, comparator)
          arr
        end

        # Return a sorted copy of the array.
        def sort(arr, &comparator)
          sort!(arr.dup, &comparator)
        end

        private

        def quick_sort(arr, low, high, comparator)
          return if low >= high

          pivot_index = partition(arr, low, high, comparator)
          quick_sort(arr, low, pivot_index - 1, comparator)
          quick_sort(arr, pivot_index + 1, high, comparator)
        end

        def partition(arr, low, high, comparator)
          # Median-of-three pivot selection
          mid = low + (high - low) / 2

          arr[low], arr[mid] = arr[mid], arr[low] if comparator.call(arr[mid], arr[low]) < 0
          arr[low], arr[high] = arr[high], arr[low] if comparator.call(arr[high], arr[low]) < 0
          arr[mid], arr[high] = arr[high], arr[mid] if comparator.call(arr[mid], arr[high]) < 0

          pivot = arr[high]
          i = low - 1

          (low...high).each do |j|
            if comparator.call(arr[j], pivot) <= 0
              i += 1
              arr[i], arr[j] = arr[j], arr[i]
            end
          end

          arr[i + 1], arr[high] = arr[high], arr[i + 1]
          i + 1
        end
      end
    end
  end
end
